File: February 2023/Day 15/Geeks Island/geeks-island.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution{   
public:
    int water_flow(vector<vector<int>> &mat,int N,int M){
    // Write your code here.
        
        vector<vector<int>> pacific_dq;
        vector<vector<int>> atlantic_dq;
        
        for(int i=0;i<N;i++) {
            pacific_dq.push_back({i,0});
            atlantic_dq.push_back({i,M-1});
        }
        
        for(int i=0;i<M;i++) {
            pacific_dq.push_back({0,i});
            atlantic_dq.push_back({N-1,i});
        }
        
        
        vector<vector<int>> is_connected_pacific = bfs(pacific_dq,mat);
        vector<vector<int>> is_connected_atlantic = bfs(atlantic_dq,mat);
        
        vector<vector<int>> ans;
        
        for(int i=0;i<N;i++) {
            for(int j=0;j<M;j++) {
                if(is_connected_pacific[i][j]==1 && is_connected_atlantic[i][j]==1) {
                    // vector<int> index{i,j};
                    ans.push_back({i,j});
                }
            }
        }
        return ans.size();
    }
    
    vector<vector<int>> bfs(vector<vector<int>>& dq,vector<vector<int>>& heights) {
        vector<vector<int>> visit(heights.size(),vector<int>(heights[0].size(),0));
        
        while(dq.size()>0) {
            vector<int> index = dq.back();
            dq.pop_back();
            int i = index[0];
            int j = index[1];
            visit[i][j]=1;
            
            // north or up
            if(isValid(visit,i-1,j) && heights[i-1][j]>=heights[i][j]) {
                dq.push_back({i-1,j});
            }
            
            // south or down
            if(isValid(visit,i+1,j) && heights[i+1][j]>=heights[i][j]) {
                dq.push_back({i+1,j});
            }
            
            // east or left
            if(isValid(visit,i,j-1) && heights[i][j-1]>=heights[i][j]) {
                dq.push_back({i,j-1});
            }
            
            // west or right
            if(isValid(visit,i,j+1) && heights[i][j+1]>=heights[i][j]) {
                dq.push_back({i,j+1});
            }
        }
        return visit;
    }
    
    bool isValid(vector<vector<int>>& visit,int i,int j) {
        if(i<0 || j<0 || i>=visit.size() || j>=visit[0].size() || visit[i][j]==1) return false;
        return true;
    }
};



//{ Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int n,m;
        cin>>n>>m;
        vector<vector<int>> mat(n, vector<int>(m));
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                cin>>mat[i][j];
            }
        }
        Solution ob;
        cout << ob.water_flow(mat, n, m) << endl;
        
    }
}


// } Driver Code Ends

File: February 2023/Day 14/Geeks Island/geeks-island.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.io.*;
import java.lang.*;

class GFG{
	public static void main(String [] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		PrintWriter out=new PrintWriter(System.out);
		int test = Integer.parseInt(br.readLine());
		while(test-- > 0) {
			String [] str = br.readLine().trim().split(" ");
			int N = Integer.parseInt(str[0]);
			int M = Integer.parseInt(str[1]);
			int [][] mat = new int[N][M];
			for(int i = 0; i < N; i++) {
				str = br.readLine().trim().split(" ");
				for(int j = 0; j < M; j++) {
					mat[i][j] = Integer.parseInt(str[j]);
				}
			}
			Solution obj = new Solution();
			out.println(obj.water_flow(mat, N, M));
		}
		out.close();
	}
}
// } Driver Code Ends


//User function Template for Java

class Solution{
	int water_flow(int [][] mat, int N, int M) {
		//Write your code here
		ArrayDeque<int[]> pacific_dq = new ArrayDeque<>();
        ArrayDeque<int[]> atlantic_dq = new ArrayDeque<>();
        
        for(int i=0;i<mat.length;i++) {
            pacific_dq.add(new int[]{i,0});
            atlantic_dq.add(new int[]{i,mat[i].length-1});
        }
        
        for(int i=0;i<mat[0].length;i++) {
            pacific_dq.add(new int[]{0,i});
            atlantic_dq.add(new int[]{mat.length-1,i});
        }
        
        
        byte[][] is_connected_pacific = bfs(pacific_dq,mat);
        byte[][] is_connected_atlantic = bfs(atlantic_dq,mat);
        
        List<List<Integer>> ans = new ArrayList<>();
        
        for(int i=0;i<mat.length;i++) {
            for(int j=0;j<mat[i].length;j++) {
                if(is_connected_pacific[i][j]==1 && is_connected_atlantic[i][j]==1) {
                    ans.add(Arrays.asList(i,j));
                }
            }
        }
        return ans.size();
	}
	
	public byte[][] bfs(ArrayDeque<int[]> dq,int[][] heights) {
        byte[][] visit = new byte[heights.length][heights[0].length];
        while(dq.size()>0) {
            int[] index = dq.pop();
            int i = index[0];
            int j = index[1];
            visit[i][j]=1;
            
            // north or up
            if(isValid(visit,i-1,j) && heights[i-1][j]>=heights[i][j]) {
                dq.add(new int[]{i-1,j});
            }
            
            // south or down
            if(isValid(visit,i+1,j) && heights[i+1][j]>=heights[i][j]) {
                dq.add(new int[]{i+1,j});
            }
            
            // east or left
            if(isValid(visit,i,j-1) && heights[i][j-1]>=heights[i][j]) {
                dq.add(new int[]{i,j-1});
            }
            
            // west or right
            if(isValid(visit,i,j+1) && heights[i][j+1]>=heights[i][j]) {
                dq.add(new int[]{i,j+1});
            }
        }
        return visit;
    }
    
    public boolean isValid(byte[][] visit,int i,int j) {
        if(i<0 || j<0 || i>=visit.length || j>=visit[0].length || visit[i][j]==1) return false;
        return true;
    }
}